### PR TITLE
[Flink-training][Gelly-task] added a new exercise which uses gelly AP…

### PR DESCRIPTION
**Pull request by samk3211**
…I. Weighted edges from the extended version of the reply-to-exercise (no duplicates) are used to generate a graph which is then given as input to a simple PageRankAlgorithm.

### CODE REVIEW

**Comment by fhueske** [comment]
Hi @samk3211, thanks for the PR! :-)
I will check it out in the next days.

File: web/exercises/replyGraphGelly.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+**Note**, the `Reply-To` field might have the value `"null"` indicating that this mail was not written in repsonse to another mail.
+
+### Expected Output
+
+
+The result for the first part of the exercise should be a `DataSet<Tuple3<String, String, Integer>>`. The first field is the sender email address of the reply mail, the second field is the sender email address of the mail that was replied to, and the third field is the number of reply connections between these two email addresses. When printed, the data set should look like this:
+
+~~~
+(fhueske@gmail.com,hongbosoftware@163.com,1)
```
Comment (fhueske) [inline web/exercises/replyGraphGelly.md]: Please use only email addresses of Flink committers (preferable only dataArtisans employees) for the training material. We don't want somebody else to be surprised to find their address here.

**Comment by fhueske** [comment]
Thanks for the PR. I see the exercise includes the extraction of the reply graph. This has been done for another exercise. I would not duplicate that code & description, but rather refer to the other exercise as a prerequisite for this one.

Also, I would not duplicate code from Flink and copy the Flink examples (EmailProfilesData, PageRankWithEdgeWeights) but only the code that a participant of the training needs to write to solve the exercise. Please have a look at the style of the other exercises.

**Comment by samk3211** [comment]
Ok,I will consider the addresses issue.
EmailProfilesData and PageRankWithEdgeWeights are part of the solution. PageRankWithEdgeWeights is the class that one needs to write in order to create a graph in gelly using the data extracted from reply graph. PageRankAlgorithm, present in Flink examples, is called through it.
EmailProfilesData has been used just to follow the pattern of the other examples in Flink: if no input file is provided then use random data. If you want me to delete that part then I can omit it.
Thanks for your feedback:)

**Comment by samk3211** [comment]
P.S.  PageRankWithEdgeWeights is very similar to the PageRank class which is already present in gelly examples. However, it is not a duplicate. In this case, edges have a weight, data input method is different and there's an additional function. These changes were required for solving this specific exercise.

I will update the excercise description to make this clearer and implement the changes suggested by you.

**Comment by fhueske** [comment]
Ah OK. Sorry, I did not take a closer look and just noticed that the structure of the reference solution looked like the examples in the Flink code base. The purpose of the built-in data is to make the examples self-contained, but blows up the reference solution quite a bit.

I would also skip the Email deduplication. That would be required to get meaningful results, but might be a bit too much for an exercise that should be solvable in about 45 mins.

**Comment by samk3211** [comment]
Ok. In this ercerice, I will leave only the graph part and omit both the duplication case and the EmailProfileData class; was just trying to follow the structure of the examples in the Flink code base. Thanks!:)

**Comment by samk3211** [comment]
Hi! I have updated the PR. If you want any further changes please let me know. Thanks :)

File: web/exercises/replyGraphGelly.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+### Input Data
+
+This exercise uses the output of the [Reply Graph]({{ site.baseurl }}/exercises/replyGraph.html) exercise as input (`CsvInputFormat`).
+
+The input data can be read as `DataSet<Tuple3<String, String, Integer>>`. The first field is the sender email address of the reply mail, the second field is the sender email address of the mail that was replied to, and the third field is the number of reply connections between these two email addresses. When printed, the data set should look like this:
+
+~~~
+(fhueske@gmail.com,hongbosoftware@163.com,1)
```
Comment (fhueske) [inline web/exercises/replyGraphGelly.md]: Can you update the example email addresses to only contain dataArtisans employees (http://data-artisans.com/)?

File: web/exercises/replyGraphGelly.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+cos@apache.org	0.001387967316018071
+~~~
+
+### Implementation Hints
+
+#### Program Structure
+
+This exercise can be solved in three steps.
+
+1. Read each tuple `DataSet<Tuple3<String, String, Double>>` of the replyGraph exercise as an Edge.
```
Comment (fhueske) [inline web/exercises/replyGraphGelly.md:41]: You list three steps here but explain only two steps later ("Generating a graph with weighted edges", "Calculating page rank").

**Comment by fhueske** [comment]
Hi @samk3211 looks pretty good :-)
Can you also add the example to the exercises overview (exercises.md) and menu (header.html)?

Later we would also need to add a Gelly lecture, but that can be done as a separate PR.

Thank you, Fabian

**Comment by samk3211** [comment]
Hi @fhueske ! 
I have done a new commit with the appropriate changes. Hope it's ok now. 
Thanks :)

**Comment by fhueske** [comment]
Found only two small things that need to be updated. Otherwise good to merge IMO.
Thank you!

**Comment by samk3211** [comment]
Done the renaming. Thanks for your feedback!

**Comment by fhueske** [comment]
Hi @samk3211,
I just merged your PR and updated the website. It's online now. 

Thank you very much for this contribution!

**Comment by vasia** [comment]
w00t! I'm very glad to see this merged :)) Thank you @samk3211, @fhueske ^^

**Pull request closed without merging**